### PR TITLE
make loginProperty public

### PR DESCRIPTION
**Pull request by ryanesch**

### CODE REVIEW

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/l10n/metatype.properties [hunk 1]
```diff
@@ -78,6 +78,9 @@ sslRef=SSL reference
 sslRef$Ref=SSL reference
 sslRef.desc=ID of the SSL configuration to be used to connect to the SSL-enabled LDAP server.
 
+loginProperty=Login property
+loginProperty.desc=The name of the login property
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/l10n/metatype.properties:82]: Can you add a period after "he name of the login property" to match the style of the other descriptions in the file?

**Review by Charlotte-Holt — COMMENTED** [review]
@ryanesch Just one ID review update required. Thx!

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/l10n/metatype.properties [hunk 2]
```diff
@@ -78,6 +78,9 @@ sslRef=SSL reference
 sslRef$Ref=SSL reference
 sslRef.desc=ID of the SSL configuration to be used to connect to the SSL-enabled LDAP server.
 
+loginProperty=Login property
+loginProperty.desc=A WIM PersonAccount property that is used to generate the LDAP filter when doing user searches. The first instance of this attribute will be returned as the principal name for the user. The mapping of WIM properties to LDAP attributes can be modified by using the attributeConfiguration attribute.
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/l10n/metatype.properties:82]: Suggested update:

>A WIM PersonAccount property that is used to generate the LDAP filter for user searches. The first instance of this attribute is returned as the principal name for the user. The mapping of WIM properties to LDAP attributes can be modified by using the attributeConfiguration attribute.

**Review by Charlotte-Holt — COMMENTED** [review]
@ryanesch Another ID review update. Let me know if you have any questions. Thx!

**Comment by andrea-pichler** [comment]
No message review requested from L2

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java [hunk 3]
```diff
@@ -634,10 +640,13 @@ private void setFilters(Map<String, Object> configProps) throws WIMSystemExcepti
 
                 // Set the login property.
                 // Remove the default uid login Property if userFilter is configured.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java:642]: This comment still valid?

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java [hunk 4]
```diff
@@ -3187,4 +3198,8 @@ public boolean isRacf() {
     public Map<String, LdapAttribute> getAttributes() {
         return iAttrNameToAttrMap;
     }
+
+    public boolean loginPropertyDefined() {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java:3202]: javadoc

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+/**
+ * Test the loginProperty. A user should be able to login with any loginProperty -
+ * with or without the userFilter defined.
+ */
+@RunWith(FATRunner.class)
+@Mode(TestMode.FULL)
+public class LoginPropertyTest {
+    //Empty server.xml. We create the config here.
+    private static LibertyServer libertyServer = LibertyServerFactory.getLibertyServer("com.ibm.ws.security.wim.adapter.ldap.fat.input.output.mapping");
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:51]: Maybe create its own server instead? Easier to find trace post run when servers aren't sharing servers.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+        if (addFilters) {
+            ldap.setCustomFilters(new LdapFilters("(&(cn=%v)(objectclass=inetorgperson))", null, "user:postalCode", null, null));
+        }
+        server.getLdapRegistries().add(ldap);
+        updateConfigDynamically(libertyServer, server);
+    }
+
+    /**
+     * Log in the with the user's uid. No loginProperties. No userFilter.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:198]: Maybe say this should be using the LDAP registries default user filters for the LDAP server type.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+     */
+    @Test
+    public void testNoLoginPropertyWithFilter() throws Exception {
+        updateLibertyServer(true);
+        assertDNsEqual("Authentication should succeed.",
+                       USER_DN, servlet.checkPassword(userCn, "password"));
+    }
+
+    /**
+     * Log in the with the user's sn from loginProperties. Include userFilter.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:222]: The userFilter should be ignored if a loginProperty is defined. I would expect this test to verify that. There should be a warning issued noting that we are ignoring the userFilter if it is also defined explicitly.

**Review by jvanhill — CHANGES_REQUESTED** [review]
I think we need a bit more testing. Personally I would do these tests (some you already have in some form).

Setup: Configure user with three attributes (sn,cn,uid perhaps).

1) Test default configuration. No explicity defined user filter or login property. Test that the user can login with UID, but not SN or CN.
2) Configure user filter using CN, but do NOT configure login property. Test that user can login with CN. This asserts that explicity defined user filter works.
3) Configure login property to be SN and CN. Test that user can login with SN and CN but NOT with UID. This verifies that the default user filter is ignored when login properties are defined. I would also test that the principal name from WIM comes back as the first login property. Maybe this could be another test.
4) Configure both user filter and login properties. This could be similar to 3, but this time a warning should be issued noting that the explicitly configured user filter is going to be ignored since login properties were defined. We will generate a dynamic filter.
5) We should test defining properties that do not exist on PersonAccount. An error should be issued.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java [hunk 8]
```diff
@@ -534,27 +534,32 @@ public Root login(Root root) throws WIMException {
...
+                }
+                Filter userFilter = iLdapConfigMgr.getUserFilter();
+                if (principalNameDN == null && userFilter != null) {
+                    sFilter = userFilter.prepare(principalName);
+                    sFilter = setAttributeNamesInFilter(sFilter, SchemaConstants.DO_PERSON_ACCOUNT);
+                }
+            }
+            if (sFilter == null) {
+                sFilter = srchCtrl.getFilter();
+                if (iLdapConfigMgr.getUseEncodingInSearchExpression() != null) {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:560]: Does this not apply if userFilter is used instead?

Comment (jvanhill) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:560]: (I'm nots sure what this call does)

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java [hunk 9]
```diff
@@ -534,27 +534,32 @@ public Root login(Root root) throws WIMException {
...
-            if (principalNameDN == null && userFilter != null) {
-                sFilter = userFilter.prepare(principalName);
-                sFilter = setAttributeNamesInFilter(sFilter, SchemaConstants.DO_PERSON_ACCOUNT);
+            if (!iLdapConfigMgr.loginPropertyDefined()) {
+                //Check is input principalName is DN, if it is DN use the same for login otherwise use userFilter
+                String principalNameDN = null;
+                try {
+                    principalNameDN = new LdapName(principalName).toString();
+                } catch (InvalidNameException e) {
+                    e.getMessage();
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:550]: Ummmm.. are we just ignoring?

Comment (ryanesch) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:550]: Yes, I didn't change that behavior, but I can remove the e.getMessage.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java (hunk 8 above)
Comment (ryanesch) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:560]: srchCtrl.getFilter does not apply if the userFilter is used. However, the userFilter is only used if loginProperties are not defined. I changed the order so that we aren't calling srchCtrl.getFilter when we aren't using it.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java (hunk 3 above)
Comment (ryanesch) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java:642]: Kind of, but I will reword it to make it more clear.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java (hunk 5 above)
Comment (ryanesch) [reply dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:51]: I agree, it's less confusing.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java (hunk 7 above)
Comment (ryanesch) [reply dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:223]: I added the warning and included a check for the warning in this test.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java (hunk 8 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:559]: I meant the getUseEncodingInSearchExpression() and encodeAttribute(). Was that a stand-alone LDAP setting? Should we be checking whether to encode attributes whether the filter is statically or dynamically defined? The answer may be "no", but I am not sure.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java [hunk 10]
```diff
@@ -534,27 +534,34 @@ public Root login(Root root) throws WIMException {
...
+                    sFilter = setAttributeNamesInFilter(sFilter, SchemaConstants.DO_PERSON_ACCOUNT);
+                }
+            }
+            if (sFilter == null) {
+                sFilter = srchCtrl.getFilter();
+                if (iLdapConfigMgr.getUseEncodingInSearchExpression() != null) {
+                    sFilter = LdapHelper.encodeAttribute(sFilter, iLdapConfigMgr.getUseEncodingInSearchExpression());
+                }
+                //Issue a warning that the filter built using loginProperty is taking precedence over the defined userFilter.
+                if (tc.isWarningEnabled())
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:563]: Lets use braces around any if statements for maintainability.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+@RunWith(FATRunner.class)
+@Mode(TestMode.FULL)
+public class LoginPropertyTest {
+    //Empty server.xml. We create the config here.
+    private static LibertyServer libertyServer = LibertyServerFactory.getLibertyServer("com.ibm.ws.security.wim.adapter.ldap.fat.loginproperty");
+    private static final Class<?> c = LoginPropertyTest.class;
+    private static UserRegistryServletConnection servlet;
+
+    /**
+     * Nearly empty server configuration. This should just contain the feature manager configuration with no
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:57]: I would just call this original configuration, and update the comment. This statement was copied from one of my original dynamic tests, and I now regret that I didn't have more generic descriptions.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+     * Setup the Liberty server. This server will start with very basic configuration. The tests
+     * will configure the server dynamically.
+     *
+     * @throws Exception If there was an issue setting up the Liberty server.
+     */
+    private static void setupLibertyServer() throws Exception {
+        /*
+         * Add LDAP variables to bootstrap properties file
+         */
+        LDAPUtils.addLDAPVariables(libertyServer);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:112]: Don't probably need this.
```suggestion
```

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        Log.info(c, "setUp", "Creating servlet connection the server");
+        servlet = new UserRegistryServletConnection(libertyServer.getHostname(), libertyServer.getHttpDefaultPort());
+
+        if (servlet.getRealm() == null) {
+            Thread.sleep(5000);
+            servlet.getRealm();
+        }
+
+        /*
+         * The original server configuration has no registry or Federated Repository configuration.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:137]:
```suggestion
         * Backup the original server configuration.
```

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+    public void testDefaultConfig() throws Exception {
+        updateLibertyServer(false);
+        assertDNsEqual("Authentication should succeed.",
+                       USER_DN, servlet.checkPassword(userUid, "password"));
+        assertNull("Authentication should not succeed.", servlet.checkPassword(userCn, "password"));
+        assertEquals(userUid, servlet.getUserDisplayName(USER_DN));
+    }
+
+    /**
+     * Log in the with the user's cn from userFilter. No loginProperties.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:216]: I would say that we define a user filter which allows a user to login with their CN to clarify the test.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+     */
+    @Test
+    public void testNoLoginPropertyWithFilter() throws Exception {
+        updateLibertyServer(true);
+        assertDNsEqual("Authentication should succeed.",
+                       USER_DN, servlet.checkPassword(userCn, "password"));
+    }
+
+    /**
+     * Log in the with the user's sn from loginProperties. Include userFilter.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:228]: Maybe say this test defines SN as a login property. It also define the userFilter. The user should be able to login with SN, but not CN and a warning should be thrown.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+     * takse precedence.
+     *
+     * @throws Exception If there was an unexpected exception.
+     */
+    @Test
+    public void testLoginPropertyWithFilter() throws Exception {
+        updateLibertyServer(true, "sn");
+        assertDNsEqual("Authentication should succeed.",
+                       USER_DN, servlet.checkPassword(userSn, "password"));
+        assertNotNull("Should find CWIML4505W message.", libertyServer.waitForStringInLog("CWIML4505W"));
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:241]: Add test before this to ensure the user can't login with CN. This way you verify that the loginProperty effectively cancels the userFilter.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+        assertEquals(userSn, servlet.getUserDisplayName(USER_DN));
+    }
+
+    /**
+     * Log in the with the user's sn from loginProperties.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:246]: I would probably just make this the same as the one above (use CN), and the key difference is NO userFilter was defined. It should perform identical to the test above I believe.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+     */
+    @Test
+    public void testLoginProperty() throws Exception {
+        updateLibertyServer(false, "sn");
+        assertDNsEqual("Authentication should succeed.",
+                       USER_DN, servlet.checkPassword(userSn, "password"));
+    }
+
+    /**
+     * Log in the with the user's sn and uid from loginProperties.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:259]: This one is using SN and CN, not UID. Maybe say that this one defines multiple login properties and NO user filter. I think all these test headers should kinda be documented the same:

Configuration:
- UserFilter: not explicitly configured
- LoginProperties: cn, sn

Expected Results
- User can login with CN and SN, but not with UID (which is defined in the default user filter).

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+
+    /**
+     * Attempt login the with the user's uid. The loginProperty
+     * defined is sn so this test should fail.
+     * This test uses the dynamically created user filter.
+     *
+     * @throws Exception If there was an unexpected exception.
+     */
+    @Test
+    public void testNonLoginProperty() throws Exception {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:286]: I think the negative test should be done for each attribute in all the other tests. SO we test login for UID, SN and CN in each test verifying that the positive and negative tests pass. Then you can remove this one.

**Review by jvanhill — CHANGES_REQUESTED** [review]

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java (hunk 8 above)
Comment (ryanesch) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapAdapter.java:559]: As discussed in slack, this is a property we have not exposed. I will handle this in an additional work item.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java (hunk 12 above)
Comment (ryanesch) [reply dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:112]: true

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java [hunk 20]
```diff
@@ -1031,20 +1040,39 @@ private void setGroupMemberFilter() {
...
+                for (Map<String, Object> propDO : loginProps) {
+                    String propName = (String) propDO.get(ConfigConstants.CONFIG_PROP_NAME);
+                    // Check if the property is supported or not
+                    List<String> props = new ArrayList<String>(1);
+                    props.add(propName);
+                    List<String> supportedProp = getSupportedProperties(acct, props);
+                    if (supportedProp == null || supportedProp.size() == 0) {
+                        unsupportedProps.add(propName);
+                    }
                     iLoginAttrs.add(getAttributeName(acct, propName));
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConfigManager.java:1062]: Don't add it if it is not supported. I realize this will have no effect at this time, b/c we throw an exception below, but if we would change the flow for some reason to tolerate bad properties, this would make sure we don't add unsupported ones.

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops [hunk 21]
```diff
@@ -221,3 +221,11 @@ LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, r
...
 LDAP_REGISTRY_INVALID_MAPPING.explanation=The custom X.509 certificate mapper must return a non-empty and non-null value or throw one of the named exceptions.
 LDAP_REGISTRY_INVALID_MAPPING.useraction=Modify the custom X.509 certificate mapper implementation so that it does not return an empty or null value. If a valid value cannot be returned, the implementation must throw a checked exception. 
+
+LOGINPROPERTY_OVERRIDE_USERFILTER=CWIML4505W: The defined userFilter attribute will be ignored since there are loginProperty attributes defined. The user filter will be dynamically generated.
+LOGINPROPERTY_OVERRIDE_USERFILTER.explanation=One or more instances of loginProperty are defined. These will be used to build the user filter.
+LOGINPROPERTY_OVERRIDE_USERFILTER.useraction=If the userFilter is preferred, remove any loginProperty attributes.
+
+INVALID_LOGIN_PROPERTIES=CWIML4506E: The following login properties are not valid: {0}.
+INVALID_LOGIN_PROPERTIES.explanation=One or more loginProperty is not a valid WIM property. These properties need to be defined in attributeConfiguration for PersonAccount.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops:230]: Make changes as we had talked about in slack.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    /**
+     * Define the loginProperty with a "bad" attribute (not a WIM property)
+     *
+     * @throws Exception If there was an unexpected exception.
+     */
+    @Test
+    public void testBadLoginProperty() throws Exception {
+        updateLibertyServer(false, "badconfig");
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:326]:
```suggestion
        updateLibertyServer(false, "nonexistentprop");
```

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 23]
```diff
@@ -0,0 +1,329 @@
...
+        assertDNsEqual("Authentication should succeed.",
+                       USER_DN, servlet.checkPassword(userSn, "password"));
+        assertDNsEqual("Authentication should succeed.",
+                       USER_DN, servlet.checkPassword(userCn, "password"));
+        assertNull("Authentication should not succeed.", servlet.checkPassword(userUid, "password"));
+        assertEquals(userCn, servlet.getUserDisplayName(USER_DN));
+    }
+
+    /**
+     * Define the loginProperty with a "bad" attribute (not a WIM property)
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:320]:
```suggestion
     * Define the loginProperty with a "bad" attribute (not a WIM PersonAccount property)
```

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 24]
```diff
@@ -0,0 +1,329 @@
...
+        ldap.setRealm("LDAPRealm");
+        ldap.setHost("localhost");
+        ldap.setPort(String.valueOf(ds.getLdapPort()));
+        ldap.setBaseDN(DN);
+        ldap.setBindDN(InMemoryLDAPServer.getBindDN());
+        ldap.setBindPassword(InMemoryLDAPServer.getBindPassword());
+        ldap.setLdapType("Custom");
+        ldap.setLdapCache(new LdapCache(new AttributesCache(false, 0, 0, "0s"), new SearchResultsCache(false, 0, 0, "0s")));
+        ldap.setAttributeConfiguration(new AttributeConfiguration());
+        //    ldap.getAttributeConfiguration().getAttributes().add(new Attribute("givenName", "badconfig", "PersonAccount", null, null));
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:187]: I see you tested if you could fix it with this... it shouldn't fix it. You should need to create an extendedProperty on PersonAccount of the same name to fix it. Perhaps in the last test where you test a bad property, you should resolve it by adding an extended property.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 25]
```diff
@@ -0,0 +1,329 @@
...
+
+        ldap.setRealm("LDAPRealm");
+        ldap.setHost("localhost");
+        ldap.setPort(String.valueOf(ds.getLdapPort()));
+        ldap.setBaseDN(DN);
+        ldap.setBindDN(InMemoryLDAPServer.getBindDN());
+        ldap.setBindPassword(InMemoryLDAPServer.getBindPassword());
+        ldap.setLdapType("Custom");
+        ldap.setLdapCache(new LdapCache(new AttributesCache(false, 0, 0, "0s"), new SearchResultsCache(false, 0, 0, "0s")));
+        ldap.setAttributeConfiguration(new AttributeConfiguration());
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:186]: Don't need this line, so either delete or uncomment it as you did with the line below.

```suggestion
```

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 26]
```diff
@@ -0,0 +1,329 @@
...
+     * Configuration:
+     *
+     * UserFilter: not explicitly configured
+     * LoginProperties: CN, SN
+     *
+     * Expected Results:
+     *
+     * User can login with CN and SN, but not with UID.
+     *
+     * Additionally, we expect the user display name to be cn as this is
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:303]: I assume we make assertion on display name because it is mapped to principalName in the input / output mappings for display name? I am not sure if that is the case or not. This would apply to all tests.

If this is the case, explain that where we make this statement.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java [hunk 27]
```diff
@@ -0,0 +1,329 @@
...
+
+    /**
+     * Define the loginProperty with a "bad" attribute (not a WIM property)
+     *
+     * @throws Exception If there was an unexpected exception.
+     */
+    @Test
+    public void testBadLoginProperty() throws Exception {
+        updateLibertyServer(false, "badconfig");
+        assertNotNull("Should find CWIML4505W message.", libertyServer.waitForStringInLog("CWIML4506E"));
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LoginPropertyTest.java:327]: As I mentioned above, maybe have a test where you resolve it by creating the extended property on PersonAccount.

**Review by jvanhill — CHANGES_REQUESTED** [review]

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/l10n/metatype.properties [hunk 28]
```diff
@@ -78,6 +78,10 @@ sslRef=SSL reference
 sslRef$Ref=SSL reference
 sslRef.desc=ID of the SSL configuration to be used to connect to the SSL-enabled LDAP server.
 
+ldap.loginProperty=Login Property
+loginProperty=Login property
+loginProperty.desc=A WIM PersonAccount property that is used to generate the LDAP filter for user searches. The first instance of this attribute is returned as the principal name for the user. The mapping of WIM properties to LDAP attributes can be modified by using the attributeConfiguration attribute. Setting this attribute will override userFilter if it is defined.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/l10n/metatype.properties:83]: Maybe we should say that these are case sensitive.

**Review by jvanhill — APPROVED** [review]
I am approving, but you should probably add a comment to the loginProperty description to say these are case-sensitive.

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
@Charlotte-Holt Hi Charlotte, can you do another review? I've updated my message and added a new one. Thanks!

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops [hunk 29]
```diff
@@ -221,3 +221,11 @@ LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, r
 LDAP_REGISTRY_INVALID_MAPPING=CWIML4504W: The custom X.509 certificate mapper returned an empty or null value. 
 LDAP_REGISTRY_INVALID_MAPPING.explanation=The custom X.509 certificate mapper must return a non-empty and non-null value or throw one of the named exceptions.
 LDAP_REGISTRY_INVALID_MAPPING.useraction=Modify the custom X.509 certificate mapper implementation so that it does not return an empty or null value. If a valid value cannot be returned, the implementation must throw a checked exception. 
+
+LOGINPROPERTY_OVERRIDE_USERFILTER=CWIML4505W: The defined userFilter attribute will be ignored since there are loginProperty attributes defined. The user filter will be dynamically generated.
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops:225]: Suggested update

>The defined userFilter attribute is ignored because loginProperty attributes are defined. The user filter is dynamically generated.

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops [hunk 30]
```diff
@@ -221,3 +221,11 @@ LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, r
...
+LOGINPROPERTY_OVERRIDE_USERFILTER.explanation=One or more instances of loginProperty are defined. These will be used to build the user filter.
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops:226]: Suggested update

>One or more instances of the loginProperty attribute are defined. These attributes are used to build the user filter.

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops [hunk 31]
```diff
@@ -221,3 +221,11 @@ LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, r
...
+LOGINPROPERTY_OVERRIDE_USERFILTER.useraction=If the userFilter is preferred, remove any loginProperty attributes.
```
Comment (Charlotte-Holt) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops:227]: Suggested update

>If the userFilter attribute is preferred, remove any loginProperty attributes.

**Review by Charlotte-Holt — COMMENTED** [review]
@ryanesch Just a few quick suggestions. Let me know if you have any questions. Thx!

**Comment by ryanesch** [comment]
#run-libby-bot
